### PR TITLE
Isolate `LibGit2Sharp` usage into a dedicated class

### PR DESCRIPTION
**Pull request by premun**
This is another step in the path to untangle the `Remote` class and similar. The main goal is to stop using the `LibGit2Sharp` library in the VMR code so that the new service can have a lower memory footprint.

The largest part is the removal of all `LibGit2Sharp` logic from `LocalGitClient` and moving it into a separate class `LocalLibGit2Client` - we can then more clearly decide when we do not want to use the `LibGit2Sharp` library.
The non-libgit2sharp code stays in a class `LocalGitClient` from which the new class inherits (so it's the old one + all libgit2sharp-enabled functionality).

### Functional changes:
- Removes the usage of `LibGit2Sharp` from all VMR code apart from the push command where it doesn't matter.
  - Adding a git remote to local repo doesn't use `LibGit2Sharp` anymore.
  - VMR synchronization does not use `LibGit2Sharp` to walk lists of commits and when listing staged files.
  - `AddRemoteIfNotExists` no longer fetches code as it was a mix of responsibilities inherited from the past.

### Other refactoring and non-functional improvements:
- Merges the `LibGit2SharpHelpers` static class into the new `LocalLibGit2Client` class.
- Some classes no longer need 2 git clients, some classes are now more clear that they have a local and remote git client.
- Some classes that had a hardcode single token now have the `RemoteConfiguration` class which holds both tokens and can resolve the correct one based on URI.
- The above simplified the `GitFileManager` class a bit as it now only needs one git client.
- We no longer need `GitFileManager` as it can now get the tokens from `RemoteConfiguration` which simplifies how the VMR code can be used from outside.

https://github.com/dotnet/arcade-services/issues/2982

### Release Note Category
- [ ] Feature changes/additions 
- [ ] Bug fixes
- [x] Internal Infrastructure Improvements

### Release Note Description
Stopped using `LibGit2Sharp` for all VMR related flows

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs [hunk 1]
```diff
@@ -315,157 +120,66 @@ public async Task<string> GetGitCommitAsync(string? repoPath = null, Cancellatio
...
+        result.ThrowIfFailed($"Failed to get remotes for {repoPath}");
 
-    private string AddRemoteIfMissing(Repository repo, string repoUrl, bool skipFetch = false)
-    {
-        var remote = repo.Network.Remotes.FirstOrDefault(r => r.Url.Equals(repoUrl, StringComparison.InvariantCultureIgnoreCase));
-        string remoteName;
+        string? remoteName = null;
 
-        if (remote is not null)
+        foreach (var line in result.StandardOutput.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
```
Comment (dkurepa) [inline src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs:155]: maybe using a Regex here is a bit easier here?
something like 
`"(\S+)\s+${repoUrl.Replace("/", "\/)}"`

File: src/Microsoft.DotNet.Darc/DarcLib/LocalLibGit2Client.cs [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+                    _logger.LogDebug($"Attempting to check out {commit} in {repoPath}");
+                    SafeCheckout(localRepo, commit, checkoutOptions);
+                    if (force)
+                    {
+                        CleanRepoAndSubmodules(localRepo);
+                    }
+                }
+                catch (NotFoundException)
+                {
+                    _logger.LogWarning($"Couldn't find commit {commit} in {repoPath} locally.  Attempting fetch.");
```
Comment (dkurepa) [inline src/Microsoft.DotNet.Darc/DarcLib/LocalLibGit2Client.cs:181]: This feels a bit wrong, like we're using try/catch for something that can be done with an if/else.
Maybe we should try to see if the repo contains the commit we're looking for with something like `git branch --contains <commit-id>`

Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/LocalLibGit2Client.cs:181]: I only copied this over from the Helper class. I would not change functionality where I don't have to ideally

Comment (dkurepa) [reply src/Microsoft.DotNet.Darc/DarcLib/LocalLibGit2Client.cs:181]: Ahh I see, I guess it's better left alone for now at least

File: src/Maestro/Maestro.DataProviders/DarcRemoteFactory.cs [hunk 3]
```diff
@@ -62,28 +62,49 @@ public async Task<IRemote> GetRemoteAsync(string repoUrl, ILogger logger)
             string normalizedUrl = AzureDevOpsClient.NormalizeUrl(repoUrl);
 
             long installationId = await Context.GetInstallationId(normalizedUrl);
```
Comment (MilenaHristova) [inline src/Maestro/Maestro.DataProviders/DarcRemoteFactory.cs:64]: what is installation id?

Comment (dkurepa) [reply src/Maestro/Maestro.DataProviders/DarcRemoteFactory.cs:64]: Id of the GitHub app that does all of the GitHub stuff

Comment (premun) [reply src/Maestro/Maestro.DataProviders/DarcRemoteFactory.cs:64]: When a repository gets created, we allow/install the maestro app to control it (make changes to it). The act/event of assignment creates some item in some GH database called installation.

**Review by MilenaHristova — APPROVED** [review]

File: src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs (hunk 1 above)
Comment (dkurepa) [reply src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs:155]: I think both of them are pretty readable and easy to understand. I guess regex on it's own, without any examples is a bit harder

**Review by dkurepa — APPROVED** [review]

**Pull request merged**